### ISSUE TITLE
Annotating F# records with Bond attributes only works when adding the CLIMutable attribute

### ISSUE THREAD

**Issue by ant0nsc** [Microsoft/bond#85]
Serializing/deserializing this record fails:
[<Bond.Schema>]
type PlainRecord = 
    {
        [<Bond.Id(0us)>] 
        DoubleImmutable     : float<kg>
    }

whereas this one works:
[<Bond.Schema>]
[<CLIMutable>]
type MutableRecord = 
    {
        [<Bond.Id(0us)>] 
        DoubleMutable : float<kg>
    }

**Comment by jack-pappas** [comment]
I think this is by design? The ``[<CLIMutable>]`` attribute causes the F# compiler to generate slightly different code for the record (compared to what it'd generate without the attribute); with the attribute, the compiler generates ``get, set`` auto-properties for the record fields so they can be set via reflection -- the F# compiler just enforces that you can't call the setters from F# code (so the record remains "immutable" as far as F# code is concerned).

``[<CLIMutable>]`` was introduced to allow F# records to be used with serializers (e.g., Bond) or ORMs, so I don't think this is a Bond-specific issue.

**Comment by ant0nsc** [comment]
Yes, I'm aware of why that attribute exists. This issue here is more of a reminder to me and @chwarr. A solution to the problem is in code review.

**Comment by ncthbrt** [comment]
Any update on this @ant0nsc ?

**Comment by chwarr** [comment]
@ant0nsc is waiting for me to implement F# codegen as part of first-class F# support. This work has been delayed because I need to work on some internal-to-Microsoft systems that use/build Bond. No ETA for when I'll be able to pick F# back up.

**Comment by ncthbrt** [comment]
Thanks for the update. Good to know it is still on the roadmap however.

On 14 Jul 2017 03:27, "Christopher Warrington" <notifications@github.com>
wrote:

@ant0nsc <https://github.com/ant0nsc> is waiting for me to implement F#
codegen as part of first-class F# support. This work has been delayed
because I need to work on some internal-to-Microsoft systems that use/build
Bond. No ETA for when I'll be able to pick F# back up.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/bond/issues/85#issuecomment-315244408>, or mute
the thread

**Comment by brettrowberry** [comment]
Interested in this!

**Comment by chwarr** [comment]
As of September 2020 and for the foreseeable future, no new languages will be added to Bond by Microsoft.

**Comment by brettrowberry** [comment]
Thanks @chwarr, should we be moving to protobuf?

**Comment by chwarr** [comment]
@brettrowberry, I've answered in issue #312, "F# code generation from IDL".
